### PR TITLE
Improve type inference of `MixtureModel`

### PR DESCRIPTION
**Pull request by devmotion**
Fixes https://github.com/JuliaStats/Distributions.jl/issues/1307 (or at least type inference of the functions that are mentioned in the issue, there might be some additional problems not covered by the tests).

The main issue was that the `prior` field is not concretely typed (`Categorical{CT}` is not a concrete type) and some hardcoded Float64 that broke type inference with dual numbers.

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 name = "Distributions"
 uuid = "31c24e10-a181-5473-b8eb-7969acd0382f"
 authors = ["JuliaStats"]
-version = "0.24.17"
+version = "0.24.18"
```
Comment (devmotion) [inline Project.toml:4]: What is the convention in Distributions - is the different type parameter of `MixtureModel` considered a breaking change?

Comment (andreasnoack) [reply Project.toml:4]: Yes

**Review by andreasnoack — COMMENTED** [review]
These are some nice simplifications. We'd have to bump the minor version to avoid breakage.

File: Project.toml (hunk 1 above)
Comment (devmotion) [reply Project.toml]: OK, I updated the version number.

Comment (devmotion) [reply Project.toml]: But now tests fail since there exists no version of HypothesisTests that is compatible with Distributions 0.25 🤔

Comment (devmotion) [reply Project.toml]: I opened a PR that removes the test dependency on HypothesisTests to fix these test failures: https://github.com/JuliaStats/Distributions.jl/pull/1311

**Comment by matbesancon** [comment]
@devmotion I think we should merge this, but merge any additional breaking change before tagging a new version

**Comment by matbesancon** [comment]
Following a conversation re-started by @StefanKarpinski, we should probably then tag a 1.0

**Comment by devmotion** [comment]
> Following a conversation re-started by @StefanKarpinski, we should probably then tag a 1.0

I am not opposed to tagging a 1.0, but it wouldn't avoid the problem for downstream packages caused by breaking changes - we just have to bump the major version instead of the minor version.

**Comment by devmotion** [comment]
> any additional breaking change before tagging a new version

What other breaking changes are ready to be merged and released? Maybe https://github.com/JuliaStats/Distributions.jl/pull/1286, or at least the bugfixes therein that could be extracted?

**Comment by matbesancon** [comment]
#1286 is not breaking is it?

**Comment by devmotion** [comment]
It changes the type of `LocationScale`, so it is breaking in the same way as this PR.

**Comment by devmotion** [comment]
I guess it would also be a good opportunity to remove the deprecations in the upcoming breaking releases that have existed for quite some time.

**Comment by devmotion** [comment]
How should we proceed? Merge https://github.com/JuliaStats/Distributions.jl/pull/1286 and this PR and then tag 0.25?

**Comment by matbesancon** [comment]
I would prefer keeping the deprecations and removing them when tagging 1.0, so yes let's go with these two PRs

**Comment by matbesancon** [comment]
@devmotion can you merge master into your branch before we merge?

**Comment by devmotion** [comment]
Done :slightly_smiling_face:

**Pull request merged**